subprojects/core/src/main/java/org/gradle/api/internal/ConventionTask.java

```diff
@@ -24,25 +24,24 @@
 import java.util.concurrent.Callable;
 
 public abstract class ConventionTask extends DefaultTask implements IConventionAware {
-    private final ConventionMapping conventionMapping;
-
-    protected ConventionTask() {
-        conventionMapping = new ConventionAwareHelper(this, getProject().getConvention());
-    }
+    private ConventionMapping conventionMapping;
 
     public Task conventionMapping(String property, Callable<?> mapping) {
-        conventionMapping.map(property, mapping);
+        getConventionMapping().map(property, mapping);
         return this;
     }
 
     public Task conventionMapping(String property, Closure mapping) {
-        conventionMapping.map(property, mapping);
+        getConventionMapping().map(property, mapping);
         return this;
     }
 
     @Override
     @Internal
     public ConventionMapping getConventionMapping() {
+        if (conventionMapping == null) {
```

Comment:
:o: Should this be synchronized? The convention may be used at execution time, I think. In particular even if it's not clean, task `A` can reference task `B` at execution time and access its convention.

Reply:
Convention mapping itself is not thread safe, so I doubt this would make any difference as far as safety goes.